pennylane/qnode.py

```diff
@@ -61,6 +61,49 @@ def pop_jacobian_kwargs(kwargs):
     return circuit_kwargs
 
 
+def _decompose_queue(ops, device):
+    """Recursively loop through a queue and decompose
+    operations that are not supported by a device.
+
+    Args:
+        ops (List[~.Operation]): operation queue
+        device (~.Device): a PennyLane device
+    """
+    new_ops = []
+
+    for op in ops:
+        if device.supports_operation(op.name):
+            new_ops.append(op)
+        else:
+            decomposed_ops = op.decomposition(*op.params, wires=op.wires)
+            decomposition = _decompose_queue(decomposed_ops, device)
+            new_ops.extend(decomposition)
+
+    return new_ops
+
+
+def decompose_queue(ops, device):
+    """Decompose operations in a queue that are not supported by a device.
+
+    This is a wrapper function for :func:`~._decompose_queue`,
+    which raises an error if an operation or its decomposition
+    is not supported by the device.
+
+    Args:
+        ops (List[~.Operation]): operation queue
+        device (~.Device): a PennyLane device
+    """
+    new_ops = []
+
```

Comment:
What if ``decompose_queue`` is itself recursive? (and then there is no need for another function to have the wrapper around):
```python

def decompose_queue(ops, device):
    """Decompose operations in a queue that are not supported by a device.

    This is a wrapper function for :func:`~._decompose_queue`,
    which raises an error if an operation or its decomposition
    is not supported by the device.

    Args:
        ops (List[~.Operation]): operation queue
        device (~.Device): a PennyLane device
    """
    new_ops = []

    for op in ops:
        if device.supports_operation(op.name):
            new_ops.append(op)
        else:
            try:
                decomposed_ops = op.decomposition(*op.params, wires=op.wires)
            except NotImplementedError:
                raise qml.DeviceError("Gate {} not supported on device {}".format(op.name, device.short_name))

            new_ops.extend(decompose_queue(decomposed_ops, device))
    return new_ops
```

Reply:
(Slight changes in tests would be required if this is implemented, for me 4 were failing, mostly related to this PR)

Reply:
I originally had a single recursive function, but there is a drawback. Since the try-except is now part of the recursion, the user will get harder to understand error messages.

For example, if they use a `U3` gate on a device that doesn't support `U3` _or_ `PhaseShift`, they will get a `PhaseShift` is not supported error. 

This is not intuitive; having the try-except outside the recursion allows for the error message to refer to top level user specified operations, in this case '`U3` is not supported'.

Reply:
Great point! That makes sense